test/libsolidity/GasTest.cpp

```diff
@@ -0,0 +1,222 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <test/libsolidity/GasTest.h>
+#include <test/Options.h>
+#include <libdevcore/CommonIO.h>
+#include <libdevcore/JSON.h>
+#include <boost/algorithm/string.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/throw_exception.hpp>
+#include <fstream>
+#include <stdexcept>
+
+using namespace langutil;
+using namespace dev::solidity;
+using namespace dev::solidity::test;
+using namespace dev;
+using namespace std;
+namespace fs = boost::filesystem;
+using namespace boost::unit_test;
+
+
+namespace
+{
+
+u256 parseGasCost(string::iterator& _it, string::iterator _end)
+{
+	if (_it == _end || !isdigit(*_it))
+		throw runtime_error("Invalid test expectation: expected gas cost.");
+	auto begin = _it;
+	while (_it != _end && isdigit(*_it))
+		++_it;
+	return u256(std::string(begin, _it));
+}
+
+}
+
+GasTest::GasTest(string const& _filename)
+{
+	ifstream file(_filename);
+	if (!file)
+		BOOST_THROW_EXCEPTION(runtime_error("Cannot open test contract: \"" + _filename + "\"."));
+	file.exceptions(ios::badbit);
+
+	m_source = parseSourceAndSettings(file);
+
+	if (m_settings.count("optimize"))
+	{
+		m_optimise = true;
+		m_validatedSettings["optimize"] = "true";
+		m_settings.erase("optimize");
+	}
+	if (m_settings.count("optimize-yul"))
+	{
+		m_optimiseYul = true;
+		m_validatedSettings["optimize-yul"] = "true";
+		m_settings.erase("optimize-yul");
+	}
+	if (m_settings.count("optimize-runs"))
+	{
+		m_optimiseRuns = stoul(m_settings["optimize-runs"]);
+		m_validatedSettings["optimize-runs"] = m_settings["optimize-runs"];
+		m_settings.erase("optimize-runs");
+	}
+
+	parseExpectations(file);
+}
+
+void GasTest::parseExpectations(std::istream& _stream)
+{
+	std::map<std::string, std::string>* currentKind = nullptr;
+	std::string line;
+
+	while (getline(_stream, line))
+		if (boost::starts_with(line, "// creation:"))
```

Comment:
Strictly, this allows for multiple `// creation:` lines, but I guess that's fine...

Reply:
True, but yes, I think it's fine.